### DISCUSSION TITLE
`const` semantics for `typeof()` expressions

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#7526] [Language Ideas]
Often times, developers will cache `typeof` expressions into a local or global variable, thinking they're saving computation by not having to re-calculate the type.

This leads to a pit of failure, as the JIT can recognize and optimize calls such as `GetType() == typeof(string)`.

PR's such as [this](https://github.com/dotnet/runtime/pull/85557) have been submitted into places like the runtime, just to fix this caching.

We should be able to write code like this
```csharp

private const Type _stringType = typeof(string);

void M(object obj)
{
    if (obj.GetType() == _stringType) { ... }
}
```
Which could reassure developers that `typeof()` can be inlined without missing out on performance.

edit: This also has Native AOT implications.

**Comment by HaloFour** [comment]
This feels like it's asking to use `const` as the opposite of a constant, as a named expression that is inlined at each use site.  I can understand the argument to only support that for `typeof(T)`, but it's still a little mind bending to me.

You'd also need a way to expose that member, fortunately C# already has a precedent of pseudo-constants with `decimal` and it'd be simple to use an attribute to embed the value.

**Reply by HaloFour** [reply]
To expand on the pseuco-constantness of this proposal, you'd want it to work like `decimal` does, where a `readonly static` field is emitted with an attribute that indicates to compilers consuming the assembly to replace the usage of that field with the underlying literal value described in the attribute.

So:

```cs
public class C1 {
    public const Type StringType = typeof(string);
}

public class C2 {
    public void M() {
        Console.WriteLine(C.StringType);
    }
}
```

is emitted as:

```cs
public class C1 {
    [System.Runtime.CompilerServices.TypeConstant(typeof(string))]
    public static readonly Type StringType = typeof(string);
}

public class C2 {
    public void M() {
        Console.WriteLine(typeof(string));
    }
}
```

**Comment by DavidArno** [comment]
I've never understood this "*asking to use `const` as the opposite of a constant*" argument against `typeof(T)` working with `const`. The only difference between that value and a string is that the string's value is determined at compile-time, whereas the type value likely has to be computed once at runtime start-up. Beyond that, both are immutable values that are therefore truly constant for the lifetime of the application.

**Reply by HaloFour** [reply]
The request is more like a macro.  It's not asking for the `typeof(T)` value to be computed once and stored off to the side.  It's asking that all named uses of the pseudo-const be replaced with the same `typeof(T)` expression, including the underlying method call, specifically because that's what the runtime recognizes and optimizes for during generic specialization.

**Reply by RenderMichael** [reply]
Isn’t that how `const` works in general? When lowering, a `const` field is inlined and gets removed from existence

**Reply by HaloFour** [reply]
The value, yes.  Not the expression used to calculate the value.

**Reply by RenderMichael** [reply]
I guess this proposal is also about rethinking the approach to `typeof`. It’s value is always constant, and (I’m no runtime expert so correct me if i’m wrong) in .NET 8 `typeof` values are stored in the frozen non-GC heap

That fact, alongside all the runtime special-casing for type comparisons, makes `typeof` closer to a literal than an expression in my eyes

**Reply by HaloFour** [reply]
Sorta, the runtime handle itself is a constant and value which is embedded directly in the assembly, but converting that to a `System.Type` does require a call to `Type.GetTypeFromHandle`.  So it is evaluated at runtime, even if it's only done once.

That's probably a minor detail, especially with the precedent of other pseudo-constants like `decimal`, where the compiler does emit a call to `Decimal`'s constructor.

I think I have more issues with the idea of adding language features specifically around JIT implementation details rather than trying to get the JIT better at recognizing existing language patterns.

**Reply by RenderMichael** [reply]
The fact that `typeof` should be used in-line isn’t just an implementation detail for JIT optimization. Making apps trimmable requires removing caching, which has Native AOT implications.

**Comment by En3Tho** [comment]
Why not just `private Type TypeOfString => typeof(string;)` instead of `private const Type _stringType = typeof(string);`?
What is the difference?

I'm sure jit is/will be able to recognize and optimize things like `private static readonly Type SomeType = LoadTypeFromSomewhere()` but these are specifically static readonly fields. This is not possible for instance fields.

**Reply by DavidArno** [reply]
> ... What is the difference? ...

You answer that question yourself: "*I'm sure jit is/will be able to recognize and optimize things like ... but these are specifically static readonly fields. **This is not possible for instance fields**.*" (emphasis, mine).

Whilst it is possible to do eg `private static readonly Type TypeOfString => typeof(string);`, that isn't then a constant, so it cannot eg be used with pattern matching.

**Reply by HaloFour** [reply]
It couldn't be used for pattern matching either way, because pattern matching hasn't been designed to support types.  The language can do that without creating pseudo-constants.

**Comment by huoyaoyuan** [comment]
`typeof` is somehow special. It can be used in some compile time constants, like attribute data.
There's also proposal about pattern matching: #356

**Reply by KennethHoff** [reply]
Ye, I was about to write a comment about the fact that Attributes already supports `typeof` despite "only"(clearly not..) supports compile-time values... But then you already had.

**Reply by RenderMichael** [reply]
Just one more example of how `typeof()` is already `const`-like

**Reply by HaloFour** [reply]
Attributes aren't specifically limited to compile-time constants.  They are limited to what the runtime can evaluate.  In the case of `System.Type`, the metadata that the runtime knows how to interpret is the embedded `System.String` of the full type name, which the runtime will attempt to resolve at runtime when you evaluate that constant.  This is very different from how `typeof(T)` works in user code, which relies on `ldtoken` with a typeref.  This is also why attributes support arrays as components, whereas arrays can't be constants.
